collector/ilm_status.go

```diff
@@ -0,0 +1,166 @@
+)
+
+type ilmStatusMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(ilm *IlmStatusResponse, status string) float64
+	Labels func(status string) []string
+}
+
+// IlmStatusCollector information struct
+type IlmStatusCollector struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	up                              prometheus.Gauge
+	totalScrapes, jsonParseFailures prometheus.Counter
+
+	metric ilmStatusMetric
+}
+
+// NewIlmStatus defines Indices IndexIlms Prometheus metrics
+func NewIlmStatus(logger log.Logger, client *http.Client, url *url.URL) *IlmStatusCollector {
+	subsystem := "ilm"
+
+	return &IlmStatusCollector{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "up"),
+			Help: "Was the last scrape of the ElasticSearch Indices Ilms endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "scrapes_total"),
+			Help: "Current total ElasticSearch Indices Ilms scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "json_parse_failures_total"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		metric: ilmStatusMetric{
+			Type: prometheus.GaugeValue,
+			Desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, subsystem, "status"),
+				"Current status of ilm. Status can be STOPPED, RUNNING, STOPPING.",
+				ilm_statuses, nil,
+			),
+			Value: func(ilm *IlmStatusResponse, status string) float64 {
+				if ilm.OperationMode == status {
+					return 1
+				}
+				return 0
+			},
+		},
+	}
+}
+
+// Describe add Snapshots metrics descriptions
+func (im *IlmStatusCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- im.metric.Desc
+	ch <- im.up.Desc()
+	ch <- im.totalScrapes.Desc()
+	ch <- im.jsonParseFailures.Desc()
+}
+
+func (im *IlmStatusCollector) getAndParseURL(u *url.URL) (*IlmStatusResponse, error) {
+	res, err := im.client.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		_ = level.Warn(im.logger).Log("msg", "failed to read response body", "err", err)
+		return nil, err
+	}
+
+	err = res.Body.Close()
+	if err != nil {
+		_ = level.Warn(im.logger).Log("msg", "failed to close response body", "err", err)
+		return nil, err
+	}
+
+	var imr IlmStatusResponse
+	if err := json.Unmarshal(body, &imr); err != nil {
+		im.jsonParseFailures.Inc()
+		return nil, err
+	}
+
+	return &imr, nil
+}
+
+func (im *IlmStatusCollector) fetchAndDecodeIlm() (*IlmStatusResponse, error) {
```

Comment:
This seems like it doesn't add value being a separate function. It could easily be part of `getAndParseURL()`